app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler.java

```diff
@@ -0,0 +1,559 @@
+                }
+
+                // also, add item and taskRun to dataMap
+                dataMap.put("item", item);
+                dataMap.put("taskRunId", taskRunId);
+
+                // get the survey schema
+                // TODO: cache this
+                UploadSchema surveySchema = uploadSchemaService.getUploadSchema(study, "ios-survey");
+                recordBuilder.withSchemaId(surveySchema.getSchemaId());
+                recordBuilder.withSchemaRevision(surveySchema.getRevision());
+            } else {
+                // JSON data may contain more than one JSON file. However, Health Data Records stores a single map.
+                // Flatten all the JSON maps together (other than info.json). We may have duplicate keys. If there are,
+                // remember those keys and log a warning afterwards.
+                Set<String> keySet = new HashSet<>();
+                Set<String> dupKeySet = new HashSet<>();
+                Map<String, JsonNode> dataFieldMap = new HashMap<>();
+                for (Map.Entry<String, JsonNode> oneJsonFile : jsonDataMap.entrySet()) {
+                    if (oneJsonFile.getKey().equals("info.json")) {
+                        // Not info.json. Skip.
+                        continue;
+                    }
+
+                    JsonNode oneJsonFileNode = oneJsonFile.getValue();
+                    Iterator<String> fieldNameIter = oneJsonFileNode.fieldNames();
+                    while (fieldNameIter.hasNext()) {
+                        String oneFieldName = fieldNameIter.next();
+
+                        // dupe field check
+                        if (keySet.contains(oneFieldName)) {
+                            dupKeySet.add(oneFieldName);
+                        } else {
+                            keySet.add(oneFieldName);
+                        }
+
+                        // If it's a dupe, the last one wins, based on JsonNode.fieldNames() iterator.
+                        dataFieldMap.put(oneFieldName, oneJsonFileNode.get(oneFieldName));
+                    }
+                }
+
+                if (!dupKeySet.isEmpty()) {
+                    addMessageAndWarn(context, String.format("Upload ID %s contains duplicate keys (%s)", uploadId,
+                            Joiner.on(", ").join(dupKeySet)));
+                }
+
+                // cross-ref our data map with our schema list to see which schema matches
+                UploadSchema latestSchema = null;
+                for (UploadSchema oneSchema : schemaList) {
+                    Set<String> schemaKeySet = new HashSet<>();
+
+                    // strategy: assume the schema matches, unless we find a field that doesn't match
+                    boolean isMatch = true;
+                    for (UploadFieldDefinition oneFieldDef : oneSchema.getFieldDefinitions()) {
+                        String fieldName = oneFieldDef.getName();
+                        schemaKeySet.add(fieldName);
```

Comment:
You have good variable names. They are short, easy to read, and mean enough to follow the code.